### ISSUE TITLE
buildDictionaryT fails to find Num and Show dictionaries for Float or Double as superclass constraints.

### ISSUE THREAD

**Issue by conal** [ku-fpg/hermit#173]
`buildDictionaryT` fails to find `Num` and `Show` dictionaries for `Float` or `Double` as superclass constraints. (It yields unbound dictionary variables instead.) It can find `Eq` and `Read` dictionaries for `Float`, `Double`, and `Int`, and it can find `Num` and `Show` dictionaries for `Int`. The `Num` and `Show` instances for `Float` and `Double` are from `GHC.Float`, while the `Eq` and `Read` instances are from `GHC.Classes` and `GHC.Read`, respectively. The `Num` and `Show` instances for `Int` are from `GHC.Num` and `GHC.Show`, respectively. I've tried adding `import GHC.Float ()` in a module of my plugin and in the examples module given to hermit, but no luck.

Any guesses about what's going wrong here (and possibly all instances from `GHC.Float`)? Any suggestions for workarounds to try?

**Comment by RyanGlScott** [comment]
FWIW, I believe this is a GHC issue, not a HERMIT one. Haddock also [has trouble](http://hackage.haskell.org/package/base-4.8.2.0/docs/Prelude.html#t:Float) locating the `Num` and `Show` instances for `Float`/`Double` for some inconspicuous reason. ([Here's the Haddock ticket.](https://github.com/haskell/haddock/issues/372))

**Comment by conal** [comment]
Thanks for the tip, @RyanGlScott. This is the first time I've tried examples with `Double` or `Float` in a HERMIT-based plugin since upgrading from GHC 7.8.2 to 7.10.3.

**Comment by RyanGlScott** [comment]
Thank makes sense, since the issue probably arose somewhere between 7.8 and 7.10, as GHC has no trouble locating the `Num`/`Show` instances for `Float` in [these GHC 7.8-compiled docs](http://hackage.haskell.org/package/base-4.7.0.2/docs/Prelude.html#t:Float) for `Prelude`.

Unfortunately, I never did get a follow-up on that Haddock ticket almost a year later. It might be worth filing as a proper GHC bug now that we've found code outside of Haddock that breaks because of it.

**Comment by xich** [comment]
Is it because the Num/Show instances are orphan instances for Float/Double? Have you tried to look up any other orphan instances?

**Comment by conal** [comment]
@xich Sounds like a good guess to me. No, I've not tried to find any other orphans.

**Comment by xich** [comment]
I'm not sure the fact that they are orphans should mean it is impossible... the typechecker must have some way to deal with orphans, I'm just not sure what it is. It is possible we just aren't initializing the typechecker state from the modguts correctly in HERMIT/GHC/Typechecker.hs.

Figuring out how the typechecker tracks orphans and looking at that file might reveal a fix.

**Comment by conal** [comment]
Good idea! I'm poking around now in the typechecker to see what I can figure out.

**Comment by RyanGlScott** [comment]
Indeed, it doesn't appear to be the case that all orphan instances are affected. Two other modules in `base` (`GHC.Base` and `GHC.Real`) export widely used orphan instances like `Monoid [a]`, but they seem to show up just fine in Haddock. I'm not sure what' different about `GHC.Float`...

**Comment by conal** [comment]
`tcg_visible_orphan_mods` seems to be key. It's set to `mkModuleSet [mod]` in `HERMIT.GHC.Typechecker`'s `initTcFromModGuts`. In GHC's `TcRnDriver` module, I see a couple of more complex initializations:

```haskell
      tcg_visible_orphan_mods = foldl extendModuleSet
                                      (tcg_visible_orphan_mods gbl)
                                      (imp_orphs imports),
```

```haskell
     , tcg_visible_orphan_mods = mkModuleSet ic_visible_mods
```
where
```haskell
   ; ic_visible_mods <- fmap concat . forM (ic_imports icxt) $ \i ->
        case i of
            IIModule n -> getOrphans n
            IIDecl i -> getOrphans (unLoc (ideclName i))
```

**Comment by conal** [comment]
Until we figure this one out, I have a workaround in the form of a `newtype` wrapper around `Double`.

**Comment by conal** [comment]
I'm no longer using HERMIT, but I've continued to run into this problem. Finally, I think I know how to fix it. If it's still an important issue for someone else, I can help.

**Comment by RyanGlScott** [comment]
@conal: What would it take to fix it? Is this a HERMIT-specific fix, or could it work for the GHC codebase itself?

**Comment by conal** [comment]
@RyanGlScott I don't know whether there's a fix needed to GHC. In `HERMIT.GHC.Typechecker`, I think the problem is in the following:

``` haskell
                tcg_imports        = emptyImportAvails,
```

When I add the module `GHC.Float`, I'm able to find instances/dictionaries for `Num Float` etc. For HERMIT, I think you can use

``` haskell
                tcg_imports        = emptyImportAvails { imp_orphs = [ mkModule baseUnitId (mkModuleName "GHC.Float") ] },
```

where

``` haskell
import Module   ( moduleName, mkModule, mkModuleName, baseUnitId )
```

I tested this tweak in a copy of `HERMIT.GHC.Typechecker`, and it fixes the issue for me.

What I'm really doing instead is using GHC's `runTcInteractive` from `TcRnDriver` in place of HERMIT's `initTcFromModGuts`, but first tweaking the given `HscEnv` `env0` to `env`, defined as follows:

``` haskell
   imports0 = ic_imports (hsc_IC env0)
   orphNames = mkModuleName <$> ["GHC.Float"]
   env = env0 { hsc_IC = (hsc_IC env0) { ic_imports = map IIModule orphNames ++ imports0 } }
```

**Comment by RyanGlScott** [comment]
FWIW, Haddock has solved a similar issue with [this patch](https://github.com/haskell/haddock/pull/724).
